Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp

```diff
@@ -1341,6 +1341,29 @@ void NetworkStorageManager::fetchLocalStorage(CompletionHandler<void(HashMap<Web
     });
 }
 
+void NetworkStorageManager::restoreLocalStorage(HashMap<WebCore::ClientOrigin, HashMap<String, String>>&& localStorageMap, CompletionHandler<void()>&& completionHandler)
+{
+    ASSERT(RunLoop::isMain());
+    ASSERT(!m_closed);
+
+    protectedWorkQueue()->dispatch([this, protectedThis = Ref { *this }, localStorageMap = crossThreadCopy(localStorageMap), completionHandler = WTFMove(completionHandler)]() mutable {
+        assertIsCurrent(workQueue());
+
+        for (auto& [clientOrigin, storageMap] : localStorageMap) {
+            auto& localStorageManager = originStorageManager(clientOrigin, ShouldWriteOriginFile::No).localStorageManager(*m_storageAreaRegistry);
+
+            if (clientOrigin.topOrigin == clientOrigin.clientOrigin)
+                localStorageManager.setLocalStorageMap(WTFMove(storageMap), clientOrigin, protectedWorkQueue());
+            else
+                localStorageManager.setTransientLocalStorageMap(WTFMove(storageMap), clientOrigin);
+        }
+
+        RunLoop::protectedMain()->dispatch([completionHandler = WTFMove(completionHandler)] mutable {
```

Comment:
I believe you need to capture `protectedThis = WTFMove(protectedThis)` here too. We need to make sure to deref on the main thread so the NetworkStorageManager gets destroyed always on the main thread. Some of its data members are Strings so the object cannot be destroyed on random threads.

Reply:
I've fixed this and that does make sense--but I'm curious--some of the other functions in this file don't do that (for example `NetworkStorageManager::requestSpace`). How come we don't need to capture `protectedThis` there?